sunpy/visualization/animator/wcs.py

```diff
@@ -0,0 +1,241 @@
+        `~astropy.visualization.wcsaxes.WCSAxes` parameters for each world
+        coordinate. The keys of this dictionary should be a value which can be
+        looked up in ``WCSAxes.coords`` (i.e. ``em.wl`` or ``hpln``) and the
+        values should be a dict which supports the following keys, and passes
+        their values to the associated `~astropy.visualization.wcsaxes.WCSAxes`
+        methods.
+
+        * ``format_unit``: `~astropy.visualization.wcsaxes.CoordinateHelper.set_format_unit`
+        * ``major_formatter``: `~astropy.visualization.wcsaxes.CoordinateHelper.set_major_formatter`
+        * ``axislabel``: `~astropy.visualization.wcsaxes.CoordinateHelper.set_axislabel`
+        * ``grid``: `~astropy.visualization.wcsaxes.CoordinateHelper.grid` (The value should be a dict of keyword arguments to ``grid()`` or `True`).
+    ylim: `tuple` or `None`, optional
+       The yaxis limits to use when drawing a line plot, if not specified
+       defaults to the data limits.
+    ylabel: `string`, optional
+       The yaxis label to use when drawing a line plot. Setting the label on
+       the y-axis on an image plot should be done via ``coord_params``.
+
+    """
+
+    def __init__(self, data, wcs, slices, coord_params=None, ylim=None, ylabel=None, **kwargs):
+        if not isinstance(wcs, BaseLowLevelWCS):
+            raise ValueError("A WCS object should be provided that implements the astropy WCS API.")
+        if wcs.pixel_n_dim is not data.ndim:
+            raise ValueError("Dimensionality of the data and WCS object do not match.")
+        if len(slices) != wcs.pixel_n_dim:
+            raise ValueError("slices should be the same length as the number of pixel dimensions.")
+        if "x" not in slices:
+            raise ValueError("slices should contain at least 'x' to indicate the axis to plot on the x axis.")
+
+        self.plot_dimensionality = 1
+
+        image_axes = [slices[::-1].index("x")]
+        if "y" in slices:
+            image_axes.append(slices[::-1].index("y"))
+            self.plot_dimensionality = 2
+
+        if self.plot_dimensionality == 1:
+            try:
+                from astropy.visualization.wcsaxes.frame import RectangularFrame1D
+            except ImportError as e:
+                raise ImportError("Astropy 4.0 must be installed to do line plotting with WCSAxes.") from e
+
+        self.naxis = data.ndim
+        self.num_sliders = self.naxis - self.plot_dimensionality
+        self.slices_wcsaxes = list(slices)
+        self.wcs = wcs
+        self.coord_params = coord_params
+        self.ylim = ylim
+        self.ylabel = ylabel
+
+        super().__init__(data, image_axes=image_axes, axis_ranges=None, **kwargs)
+
+    def _sanitize_axis_ranges(self, *args):
+        """
+        This overrides the behaviour of ArrayAnimator to generate axis_ranges
+        based on the WCS.
+        """
+
+        def partial_pixel_to_world(pixel_dimension, pixel_coord):
```

Comment:
What's the advantage of defining a function within a function here?

Reply:
There isn't much either way really, it's just where the code belonged to me!

There is a potential performance optimisation in using closures here (for the lines in this function which are not dependant upon the pixel coord), which would require the function to be defined here, so I think it's probably a sane place for it.

Reply:
OK. So long as there's a reason to justify its ugliness ;)

Reply:
On further reflection I have moved it to its own method.